### ISSUE TITLE
get_stable_state

### ISSUE THREAD

**Issue by v-wewei** [zdchan/GraspXL#15]
the code for get_stable_state is missing ？

**Comment by zdchan** [comment]
Hi the stable states were recorded by letting the objects fall to the table without hand interaction. We didn't save the script. You can directly set the hand aside, let the objects fall, and record the final frame. Alternatively, I quickly wrote one, for which you can put under allegro_test to run. (comment out the link actionMean_r_.head(3) += wrist_bias in environment.hpp https://github.com/zdchan/GraspXL/blob/e72e0a42747565781cb308e54cb8a9aa08407c35/raisimGymTorch/raisimGymTorch/env/envs/allegro_test/Environment.hpp#L471, and rerun python setup,py develop)


`#!/usr/bin/python

from ruamel.yaml import YAML, dump, RoundTripDumper
from raisimGymTorch.env.bin import allegro_test as mano
from raisimGymTorch.env.RaisimGymVecEnvOther import RaisimGymVecEnvTest as VecEnv
from raisimGymTorch.helper.raisim_gym_helper import ConfigurationSaver, load_param, tensorboard_launcher
from raisimGymTorch.env.bin.allegro_test import NormalSampler
from random import choice
from raisimGymTorch.helper.initial_pose_final import get_initial_pose_set, get_initial_pose_faive

# import pdb

import os
import math
import time
import raisimGymTorch.algo.ppo.module as ppo_module
import raisimGymTorch.algo.ppo.ppo as PPO
import torch.nn as nn
import numpy as np
import torch
from datetime import datetime
import argparse
from raisimGymTorch.helper import rotations
import joblib
import random
import wandb
import torch
import shutil

exp_name = "allegro_floating"

# weight_saved = '2024-01-28-11-14-31/full_5800_r.pt'
# weight_saved = '2024-01-30-16-33-20/full_4400_r.pt'
# weight_saved = '2024-01-30-16-33-20/full_8000_r.pt'
weight_saved = '2024-01-30-16-35-15/full_7300_r.pt'

record = True

# configuration
parser = argparse.ArgumentParser()
parser.add_argument('-c', '--cfg', help='config file', type=str, default='cfg_reg.yaml')
parser.add_argument('-d', '--logdir', help='set dir for storing data', type=str, default=None)
parser.add_argument('-e', '--exp_name', help='exp_name', type=str, default=exp_name)
parser.add_argument('-w', '--weight', type=str, default=weight_saved)
parser.add_argument('-sd', '--storedir', type=str, default='data_all')
parser.add_argument('-seed', '--seed', type=int, default=1)
parser.add_argument('-itr', '--num_batches', type=int, default=5)
parser.add_argument('-nr', '--num_repeats', type=int, default=1)
parser.add_argument('-re', '--load_trained_policy', action="store_true")
parser.add_argument('-renew', '--renew', help='update labels every iteration', action="store_true")
parser.add_argument('-ln', '--log_name', type=str, default='single_obj')
parser.add_argument('-mean', '--mean_pose', action="store_true")


args = parser.parse_args()
weight_path = args.weight
cfg_grasp = args.cfg

print(f"Configuration file: \"{args.cfg}\"")
print(f"Experiment name: \"{args.exp_name}\"")

# task specification
task_name = args.exp_name
# check if gpu is available
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
# directories
task_path = os.path.dirname(os.path.realpath(__file__))
home_path = task_path + "/../../../../.."

if args.logdir is None:
    exp_path = home_path
else:
    exp_path = args.logdir

# config
cfg = YAML().load(open(task_path + '/cfgs/' + args.cfg, 'r'))

if args.seed != 1:
    cfg['seed'] = args.seed

cfg['environment']['visualize'] = True





# cat_name = 'affordance_level'
# cat_name = 'gen_obj'
# cat_name = 'YCB'
# cat_name = 'wild'
cat_name = 'large_scale_obj'

cfg['environment']['load_set'] = cat_name
directory_path = home_path + f"/rsc/{cat_name}/"
print(directory_path)

items = os.listdir(directory_path)

# # Filter out only the folders (directories) from the list of items
folder_names = [item for item in items if os.path.isdir(os.path.join(directory_path, item))]

obj_path_list = []
obj_ori_list = folder_names


num_env_per_iter = args.num_repeats
iter_num = args.num_batches

# num_env_per_iter = 1
# iter_num = random.randint(0, len(obj_ori_list) // num_env_per_iter + int(len(obj_ori_list) % num_env_per_iter > 0))
# for iter_num in range(len(obj_ori_list) // num_env_per_iter + int(len(obj_ori_list) % num_env_per_iter > 0)):
#     print("iter_num", iter_num)
obj_list = []
if iter_num < len(obj_ori_list) // num_env_per_iter:
    num_envs = num_env_per_iter
else:
    num_envs = len(obj_ori_list) % num_env_per_iter
for i in range(num_envs):
    obj_list.append(obj_ori_list[i + iter_num * num_env_per_iter])
print(obj_list)
print("iter_num", iter_num)


num_envs = len(obj_list)

activations = nn.LeakyReLU

cfg['environment']['num_envs'] = num_envs
print('num envs', num_envs)

# Environment definition
env = VecEnv(obj_list, mano.RaisimGymEnv(home_path + "/rsc", dump(cfg['environment'], Dumper=RoundTripDumper)),
             cfg['environment'], cat_name=cat_name)

print("initialization finished")

for obj_item in obj_list:
    obj_path_list.append(os.path.join(f"{obj_item}/{obj_item}.urdf"))
env.load_multi_articulated(obj_path_list)

ob_dim_r = 210
act_dim = 22
print('ob dim', ob_dim_r)
print('act dim', act_dim)

# Training
trail_steps = 30
reward_clip = -2.0
grasp_steps = 80
lift_step = 50
n_steps_r = grasp_steps + trail_steps + lift_step
total_steps_r = n_steps_r * env.num_envs

# RL network

actor_r = ppo_module.Actor(
    ppo_module.MLP(cfg['architecture']['policy_net'], activations, ob_dim_r, act_dim),
    ppo_module.MultivariateGaussianDiagonalCovariance(act_dim, num_envs, 1.0, NormalSampler(act_dim)), device)

critic_r = ppo_module.Critic(ppo_module.MLP(cfg['architecture']['value_net'], activations, ob_dim_r, 1), device)


test_dir = True

saver = ConfigurationSaver(log_dir=exp_path + "/raisimGymTorch/" + args.storedir + "/" + task_name,
                           save_items=[task_path + "/affordance_eval.py"], test_dir=test_dir)


ppo_r = PPO.PPO(actor=actor_r,
                critic=critic_r,
                num_envs=num_envs,
                num_transitions_per_env=n_steps_r,
                num_learning_epochs=4,
                gamma=0.996,
                lam=0.95,
                num_mini_batches=4,
                device=device,
                log_dir=saver.data_dir,
                shuffle_batch=False
                )
# load_param(saver.data_dir.split('eval')[0]+weight_path, env, actor_r, critic_r, ppo_r.optimizer, saver.data_dir, cfg_grasp)

success_rate = 0.0
ori_error = 0.0
direction_error = 0.0
rotation_error = 0.0
center_error = 0.0
contact_ratio = 0.0
suc_ori_error = 0.0
suc_direction_error = 0.0
suc_rotation_error = 0.0
suc_center_error = 0.0
suc_contact_ratio = 0.0

lifted_num = 0

for update in range(1):
    start = time.time()

    #TODO: 可以用那个数据集里的初始手势作为我们的初始手势，这样只需要调整方向
    qpos_reset_r = np.zeros((num_envs, 22), dtype='float32')
    qpos_reset_l = np.zeros((num_envs, 22), dtype='float32')
    obj_pose_reset = np.zeros((num_envs, 8), dtype='float32')

    target_center = np.zeros_like(env.affordance_center)
    object_center = np.zeros_like(env.affordance_center)
    fake_non_aff_center = [0.346408, 0.346408, 0.346408]
    contain_non_aff = np.zeros((num_envs, 1), dtype='float32')

    for i in range(num_envs):
        lowest_point = 0.
        txt_file_path = os.path.join(directory_path, obj_list[i]) + "/lowest_point_new.txt"
        with open(txt_file_path, 'r') as txt_file:
            lowest_point = float(txt_file.read())

        obj_pose_reset[i, :] = [1., -0., 0.505, 1., -0., -0., 0., 0.]
        obj_pose_reset[i, 2] -= lowest_point

        qpos_reset_r[i, -4] = 1.7


        if np.linalg.norm(env.non_aff_mesh[i].centroid - fake_non_aff_center) < 0.01:
            non_aff_mesh = None
        else:
            non_aff_mesh = env.non_aff_mesh[i]
            contain_non_aff[i, 0] = 1.


        # direction = np.array([[0., 1, 0]])
        # rotation = 3.14
        # point = np.array([0.0, 0.0, -0.0])
        # # point = env.affordance_center[i]
        # rot, pos, bias = get_initial_pose_set(env.aff_mesh[i], non_aff_mesh, direction, rotation, point, hand='allegro')
        rot, pos, bias = get_initial_pose_faive(env.aff_mesh[i], non_aff_mesh, "allegro")

        obj_mat = rotations.quat2mat(obj_pose_reset[i, 3:7])
        wrist_pose_obj = rotations.axisangle2euler(rot.reshape(-1, 3)).reshape(1, -1)
        wrist_mat = rotations.euler2mat(wrist_pose_obj)
        wrist_in_world = np.matmul(obj_mat, wrist_mat)
        wrist_pose = rotations.mat2euler(wrist_in_world)
        qpos_reset_r[i, :3] = obj_pose_reset[i, :3] + np.matmul(obj_mat, pos[0, :])
        qpos_reset_r[i, 3:6] = wrist_pose[0, :]

        target_center[i, :] = bias[:]
        object_center[i, :] = env.affordance_center[i]
    print("complete initial pose generation")


    env.reset_state(qpos_reset_r,
                    qpos_reset_l,
                    np.zeros((num_envs, 22), 'float32'),
                    np.zeros((num_envs, 22), 'float32'),
                    obj_pose_reset,
                    )

    env.set_goals(target_center,
                  object_center,
                  np.zeros((num_envs, 1), 'float32'),
                  np.zeros((num_envs, 1), 'float32'),
                  np.zeros((num_envs, 1), 'float32'),
                  np.zeros((num_envs, 1), 'float32'),
                  np.zeros((num_envs, 1), 'float32'),
                  np.zeros((num_envs, 1), 'float32'),
                  np.zeros((num_envs, 1), 'float32'),
                  np.zeros((num_envs, 1), 'float32'),
                  )

    obs_new_r, _ = env.observe(contain_non_aff, allegro=True)
    if record:
        trans_r = np.zeros((num_envs, n_steps_r, 3))
        rot_r = np.zeros((num_envs, n_steps_r, 3))
        pose_r = np.zeros((num_envs, n_steps_r, 22))
        trans_obj = np.zeros((num_envs, n_steps_r, 3))
        rot_obj = np.zeros((num_envs, n_steps_r, 3))
        angle_obj = np.zeros((num_envs, n_steps_r, 1))
    
    # Record object pose history
    obj_pose_history = np.zeros((n_steps_r, num_envs, 8), dtype='float32')
    
    for step in range(n_steps_r):
        # print("step", step)
        obs_r = obs_new_r
        obs_r = obs_r[:, :].astype('float32')
        env.switch_root_guidance(True)

        action_r = actor_r.architecture.architecture(torch.from_numpy(obs_r.astype('float32')).to(device))
        action_r = action_r.cpu().detach().numpy()
        action_l = np.zeros_like(action_r)


        frame_start = time.time()

        reward_r, _, dones = env.step(action_r.astype('float32'), action_l.astype('float32'))

        obs_new_r, dis_info = env.observe(contain_non_aff, allegro=True)
        obs_new_r = obs_new_r[:].astype('float32')
        global_state = env.get_global_state().astype('float32')

        obj_pose_history[step] = global_state[:, -8:]  # Last 8 values are the object pose

        frame_end = time.time()
        wait_time = cfg['environment']['control_dt'] - (frame_end - frame_start)
        if wait_time > 0.:
            time.sleep(wait_time)

    
    # Save object stable state
    for i in range(num_envs):
        # Check for anomalous values
        if np.any(obj_pose_history[:, i, 7] > 0.5):
            print(f"Warning: NaN or abnormal values found for {obj_item}")
        else:
            # Save the last frame pose as the stable state
            stable_state = obj_pose_history[-1, i].reshape(1, 8)
            print(f"Stable state for {obj_item}: {stable_state}")
            
            # Save stable state to the object's own directory
            obj_state_path = os.path.join(directory_path, obj_item) + "/obj_stable_state.txt"
            with open(obj_state_path, 'w') as f:
                f.write(' '.join(map(str, stable_state.flatten())))
            
            print(f"Stable state saved to: {obj_state_path}")

    print(f"Completed simulation for {obj_item}")
`

**Comment by v-wewei** [comment]
Thanks for your reply, i will test it soon.
